### ISSUE TITLE
Extend let_and_return to an arbitrary expression.

### ISSUE THREAD

**Issue by Sean1708** [Manishearth/rust-clippy#340]
In a piece of code I ended up with something similar to (excuse the contrived example)

```rust
let c = Some(10);
let b = true;
let a = if b {
    let temp = match c {
        Some(i) => i,
        None => panic!(),
    };
    temp
} else {
    -10
};
```

which is semantically equivalent to

```rust
let c = Some(10);
let b = true;
let a = if b {
    match c {
        Some(i) => i,
        None => panic!(),
    }
} else {
    -10
};
```
yet clippy didn't catch it.

By the looks of it `let_and_return` only works at the end of the a function, is it possible to make it work in any arbitrary expression?

**Comment by birkenfeld** [comment]
You are right. Opened a PR for this.

**Comment by Sean1708** [comment]
That's working great for me, thanks.
